django/test/testcases.py

```diff
@@ -1510,10 +1512,24 @@ def captureOnCommitCallbacks(cls, *, using=DEFAULT_DB_ALIAS, execute=False):
         finally:
             while True:
                 callback_count = len(connections[using].run_on_commit)
-                for _, callback in connections[using].run_on_commit[start_count:]:
+                for _, callback, is_robust in connections[using].run_on_commit[
+                    start_count:
+                ]:
                     callbacks.append(callback)
-                    if execute:
-                        callback()
+                    if is_robust:
+                        try:
+                            if execute:
+                                callback()
+                        except Exception as e:
+                            logger.error(
+                                f"Error calling {callback.__qualname__} on_commit() "
+                                f"during transaction (%s).",
+                                e,
+                                exc_info=True,
+                            )
+                    else:
+                        if execute:
+                            callback()
```

Comment:
I'd revert the logic i.e. check `execute` first:
```suggestion
                    if execute:
                        if is_robust:
                            try:
                                callback()
                            except Exception as e:
                                logger.error(
                                    f"Error calling {callback.__qualname__} on_commit() "
                                    f"during transaction (%s).",
                                    e,
                                    exc_info=True,
                                )
                        else:
                            callback()
```